File: backend/src/main/kotlin/org/gymapp/backend/mapper/UserMapper.java
package org.gymapp.backend.mapper;

import org.gymapp.backend.model.GymUser;
import org.gymapp.backend.model.User;
import org.gymapp.library.request.CreateUserRequest;
import org.gymapp.library.response.UserDto;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.Mappings;

import java.util.List;

@Mapper(componentModel = "spring")
public interface UserMapper {

    @Mapping(target = "createdAt", expression = "java(java.time.LocalDateTime.now())")
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "gymUsers", ignore = true)
    User requestToUser(CreateUserRequest createUserRequest);

    @Mapping(source = "gymUsers", target = "gymUsersIds")
    UserDto modelToDto(User user);

    List<UserDto> modelsToDtos(List<User> user); 

    default String mapGymUserToId(GymUser gymUser) {
        return gymUser.getId();
    }
}

File: backend/src/main/kotlin/org/gymapp/backend/mapper/UserMapper.kt
package org.gymapp.backend.mapper

import org.gymapp.backend.model.GymUser
import org.gymapp.backend.model.User
import org.gymapp.library.request.CreateUserRequest
import org.gymapp.library.response.UserDto
import org.mapstruct.*
import org.mapstruct.factory.Mappers

@Mapper(componentModel = "spring")
interface UserMapper {

    companion object {
        val INSTANCE: UserMapper = Mappers.getMapper(UserMapper::class.java)
    }

    @Mappings(
        Mapping(target = "createdAt", expression = "java(java.time.LocalDateTime.now())"),
        Mapping(target = "updatedAt", ignore = true),
        Mapping(target = "gymUsers", ignore = true)
    )
    fun requestToUser(createUserRequest: CreateUserRequest): User

    @Mapping(source = "gymUsers", target = "gymUsersIds")
    fun modelToDto(user: User): UserDto

    fun modelsToDtos(users: List<User>): List<UserDto>

    fun mapGymUserToId(gymUser: GymUser): String {
        return gymUser.id
    }
}